pkg/chunk/chunk_store.go

```diff
@@ -149,19 +166,44 @@ func (c *store) calculateIndexEntries(userID string, from, through model.Time, c
 	}
 	indexEntriesPerChunk.Observe(float64(len(entries)))
 
+	uncachedEntries, cacheKeys := c.dedupeEntriesFromCache(entries)
+	indexWritesTotal.Add(float64(len(uncachedEntries)))
+
 	// Remove duplicate entries based on tableName:hashValue:rangeValue
 	result := c.storage.NewWriteBatch()
-	for _, entry := range entries {
+	for _, entry := range uncachedEntries {
 		key := fmt.Sprintf("%s:%s:%x", entry.TableName, entry.HashValue, entry.RangeValue)
 		if _, ok := seenIndexEntries[key]; !ok {
 			seenIndexEntries[key] = struct{}{}
 			rowWrites.Observe(entry.HashValue, 1)
 			result.Add(entry.TableName, entry.HashValue, entry.RangeValue, entry.Value)
 		}
 	}
+	c.entryCache.Store(context.Background(), cacheKeys, make([][]byte, len(cacheKeys)))
```

Comment:
I'm a little worries that if the write fails we won't try again as you've added it to the cache too early.

Reply:
This will lose data, at least on DynamoDB where the whole write can fail and be retried.